<a href="https://colab.research.google.com/github/victoraccete/LED_tasks/blob/main/Desafio_LED_Parte_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [19]:
import pandas as pd

# Parte 1

In [20]:
us_500 = pd.read_csv("https://raw.githubusercontent.com/victoraccete/LED_tasks/main/us-500.csv")
us_500.head()

first_name  ...                                  web
0      James  ...         http://www.bentonjohnbjr.com
1  Josephine  ...     http://www.chanayjeffreyaesq.com
2        Art  ...       http://www.chemeljameslcpa.com
3      Lenna  ...  http://www.feltzprintingservice.com
4    Donette  ...    http://www.printingdimensions.com

[5 rows x 12 columns]

## A. Cheque se há algum dado faltante na base de dados ‘us-500’;
Podemos fazer essa checagem de muitas maneiras. A primeira, é usar o pandas para obter informações das colunas do conjunto de dados e ler se há alguma coluna cujo número de non-null é menor que o número total de observações. Assim:

In [21]:
us_500.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   first_name    500 non-null    object
 1   last_name     500 non-null    object
 2   company_name  500 non-null    object
 3   address       500 non-null    object
 4   city          500 non-null    object
 5   county        500 non-null    object
 6   state         500 non-null    object
 7   zip           500 non-null    int64 
 8   phone1        500 non-null    object
 9   phone2        500 non-null    object
 10  email         500 non-null    object
 11  web           500 non-null    object
dtypes: int64(1), object(11)
memory usage: 47.0+ KB


Já podemos ver que não há dados faltantes neste conjunto de dados.  
No entanto, há outras formas de fazer essa verificação. Podemos, por exemplo, somar os números de dados faltantes, por coluna, no conjunto. Como não há dados faltantes no dataset, o resultado esperado é que cada coluna seja 0. Vamos ver:

In [22]:
us_500.isna().sum()

first_name      0
last_name       0
company_name    0
address         0
city            0
county          0
state           0
zip             0
phone1          0
phone2          0
email           0
web             0
dtype: int64

## B. Classifique os endereços por ‘state’;

Confesso que não entendi exatamente o que seria essa classificação por estado. Suponho que seja um agrupamento. Se esse for o caso, podemos fazer da seguinte forma:

In [23]:
addresses = us_500[['state', 'address']].groupby(by=['state', 'address']).sum()
addresses

Empty DataFrame
Columns: []
Index: [(AK, 1048 Main St), (AK, 1747 Calle Amanecer #2), (AK, 18 Fountain St), (AK, 20 S Babcock St), (AK, 639 Main St), (AK, 735 Crawford Dr), (AR, 1844 Southern Blvd), (AZ, 1 Huntwood Ave), (AZ, 10759 Main St), (AZ, 347 Chestnut St), (AZ, 4 58th St #3519), (AZ, 72 Southern Blvd), (AZ, 73 State Road 434 E), (AZ, 86350 Roszel Rd), (AZ, 9 W Central Ave), (AZ, 96541 W Central Blvd), (CA, 1 Century Park E), (CA, 10276 Brooks St), (CA, 105 Richmond Valley Rd), (CA, 1128 Delaware St), (CA, 11360 S Halsted St), (CA, 13252 Lighthouse Ave), (CA, 17 Morena Blvd), (CA, 18 Coronado Ave #563), (CA, 1933 Packer Ave #2), (CA, 195 13n N), (CA, 2 Monroe St), (CA, 20932 Hedley St), (CA, 2140 Diamond Blvd), (CA, 2184 Worth St), (CA, 22 Spruce St #595), (CA, 2409 Alabama Rd), (CA, 25 E 75th St #69), (CA, 251 Park Ave #979), (CA, 26659 N 13th St), (CA, 2887 Knowlton St #5435), (CA, 2972 Lafayette Ave), (CA, 30 W 80th St #1995), (CA, 3196 S Rider Trl), (CA, 326 E Main St #6496), (CA, 34 Raritan Center Pky), (CA, 35433 Blake St #588), (CA, 3717 Hamann Industrial Pky), (CA, 38 Pleasant Hill Rd), (CA, 3882 W Congress St #799), (CA, 39 Moccasin Dr), (CA, 395 S 6th St #2), (CA, 3989 Portage Tr), (CA, 4 E Colonial Dr), (CA, 4 Otis St), (CA, 4 S Washington Ave), (CA, 4119 Metropolitan Dr), (CA, 429 Tiger Ln), (CA, 433 Westminster Blvd #590), (CA, 469 Outwater Ln), (CA, 47154 Whipple Ave Nw), (CA, 47939 Porter Ave), (CA, 48 Stratford Ave), (CA, 4800 Black Horse Pike), (CA, 5 N Cleveland Massillon Rd), (CA, 5 Tomahawk Dr), (CA, 5 Washington St #1), (CA, 524 Louisiana Ave Nw), (CA, 52777 Leaders Heights Rd), (CA, 577 Parade St), (CA, 598 43rd St), (CA, 6 Greenleaf Ave), (CA, 6 Middlegate Rd #106), (CA, 60 Fillmore Ave), (CA, 62 Monroe St), (CA, 68556 Central Hwy), (CA, 69 Marquette Ave), (CA, 6916 W Main St), (CA, 7 W Jackson Blvd), (CA, 70 Mechanic St), (CA, 701 S Harrison Rd), (CA, 72119 S Walker Ave #63), (CA, 72868 Blackington Ave), (CA, 77 222 Dr), (CA, 82 N Highway 67), (CA, 83649 W Belmont Ave), (CA, 87 Sierra Rd), (CA, 87895 Concord Rd), (CA, 9 N 14th St), (CA, 94290 S Buchanan St), (CA, 98 University Dr), (CA, 985 E 6th Ave), (CA, 99385 Charity St #840), (CO, 17 Jersey Ave), (CO, 2664 Lewis Rd), (CO, 4 Webbs Chapel Rd), (CO, 669 Packerland Dr #1438), (CO, 721 Interstate 45 S), (CO, 76598 Rd  I 95 #1), (CO, 8 Us Highway 22), (CO, 866 34th Ave), (CT, 46314 Route 130), (CT, 49440 Dearborn St), (CT, 759 Eldora St), (CT, 78112 Morris Ave), ...]

[500 rows x 0 columns]

Assim, agrupamos os endereços por estado. No entanto, ambos são considerados índices. Para considerar 'address' e 'state' como colunas, podemos fazer um reset_index():

In [24]:
addresses.reset_index()

state                 address
0      AK            1048 Main St
1      AK  1747 Calle Amanecer #2
2      AK          18 Fountain St
3      AK         20 S Babcock St
4      AK             639 Main St
..    ...                     ...
495    WI           762 S Main St
496    WI      9 N College Ave #3
497    WY       38773 Gravois Ave
498    WY     66697 Park Pl #3224
499    WY     7140 University Ave

[500 rows x 2 columns]

Caso por "classificação" a intenção seja no sentido de ordenação, podemos fazer o seguinte:

In [25]:
us_500.sort_values(by='state').head(8)

first_name  ...                                      web
386       Gail  ...        http://www.servicesupplycoinc.com
66      Penney  ...         http://www.hawaiiankinghotel.com
3        Lenna  ...      http://www.feltzprintingservice.com
42      Roxane  ...          http://www.rapidtradingintl.com
44       Erick  ...     http://www.cindyturnerassociates.com
93       Wilda  ...      http://www.mclaughlinlutherwcpa.com
159      Carin  ...             http://www.redekerdebbie.com
286   Herminia  ...  http://www.seaislanddivoffstrindinc.com

[8 rows x 12 columns]

## C. Suponha que o zip code deve sempre ter 5 dígitos, ache todos os zip codes irregulares;


Para contar o número de strings, podemos aplicar a função len() para cada string. Mas, para fazer isso em cada observação, devemos usar a função apply, que aplica uma função a cada observação do dataframe. Neste caso, vamos usar uma função lambda, que vai receber a entrada (que é numérica) e vamos retornar a quantidade de caracteres da string da instância. 


In [26]:
us_500['zip'].apply(lambda x: len(str(x)))

0      5
1      5
2      4
3      5
4      5
      ..
495    5
496    5
497    5
498    5
499    5
Name: zip, Length: 500, dtype: int64

Conseguimos ver acima que algumas observações possuem 4 dígitos. No entanto, ainda não conseguimos visualizar quantas observações possui mais de 5 caracteres ainda. Para isso, fazemos:

In [27]:
us_500[us_500['zip'].apply(lambda x: len(str(x))) > 5]

Empty DataFrame
Columns: [first_name, last_name, company_name, address, city, county, state, zip, phone1, phone2, email, web]
Index: []

Ou, se quisermos só o número:

In [28]:
us_500['zip'][us_500['zip'].apply(lambda x: len(str(x))) > 5].count()

0

No entanto, como 5 é o único número que aceitamos como regular, podemos pegar só as observações cujo número de dígitos do zip code é diferente de 5. 

In [29]:
irregular_zip_codes_df = us_500[us_500['zip'].apply(lambda x: len(str(x))) != 5]
irregular_zip_codes_df.head()

first_name  ...                               web
2         Art  ...    http://www.chemeljameslcpa.com
24     Alisha  ...        http://www.wtlzpowerfm.com
32      Ernie  ...      http://www.knwznewsradio.com
33     Albina  ...  http://www.giampetroanthonyd.com
46       Jina  ...   http://www.gracepastriesinc.com

[5 rows x 12 columns]

Acima, podemos ver algumas das obversações com zip code irregular. Se precisarmos, podemos pegar apenas seus índices:

In [30]:
irregular_zip_codes_df.index

Int64Index([  2,  24,  32,  33,  46,  54,  58,  60,  61,  77,  79, 108, 113,
            119, 120, 126, 127, 132, 133, 134, 135, 149, 157, 171, 177, 179,
            182, 188, 190, 193, 196, 199, 222, 231, 235, 237, 242, 248, 249,
            253, 265, 268, 275, 283, 294, 296, 306, 311, 317, 319, 321, 330,
            332, 335, 342, 364, 373, 385, 397, 403, 404, 406, 417, 419, 420,
            429, 431, 441, 442, 448, 457, 458, 463, 466, 478, 484, 488, 489],
           dtype='int64')

## D. Separe a coluna ‘address’ em três partes: o número na rua, o nome da rua e o número de quartos. (Considere que pode haver partes faltantes)

Como aconselhado, vamos levar em consideração que pode haver partes faltantes. Numa visualização rápida, podemos ver que realmente há partes faltantes:

In [31]:
us_500['address'].head()

0      6649 N Blue Gum St
1     4 B Blue Ridge Blvd
2    8 W Cerritos Ave #54
3             639 Main St
4            34 Center St
Name: address, dtype: object

Podemos visualizar rapidamente que realmente há casos em que não há algumas partes. Vamos conferir se a primeira parte, numérica, está presente em todas as observações:

In [32]:
us_500['address'].apply(lambda x: x.split()[0].isnumeric()).sum()

500

Podemos ver que a última parte, quando presente, começa com o caracter '#'. Por isso, vamos checar quantas vezes a última parte começa com '#'. 

In [33]:
us_500['address'].apply(lambda x: x.split()[-1].startswith('#')).sum()

109

Assim, vemos que apenas 109 das 500 observações possuem o número. Por isso, precisamos tratar esses casos. 

In [34]:
address_1 = '8 W Cerritos Ave #54'
address_2 = '6649 N Blue Gum St'

def split_address(address: str) -> tuple:
    ''' Takes an address string and splits the initial number, the name of the street,
    avenue, boulevard or other alike places, and the final number, following a '#'. 
    Params:
        address: the address string
    Returns: 
        a tuple containing the three parts.
    '''
    splitted = address.split()
    pt_1 = splitted.pop(0)
    if splitted[-1].startswith('#'):
        pt_3 = splitted.pop(-1)
    else:
        pt_3 = ''
    pt_2 = " ".join(splitted)

    return pt_1, pt_2, pt_3

assert split_address(address_1) == ('8', 'W Cerritos Ave', '#54')
assert split_address(address_2) == ('6649', 'N Blue Gum St', '')

A função acima serve para separar o endereço, como pedido no enunciado. Se quisermos aplicar a função para todo o dataset, basta fazer: 

In [35]:
us_500['address'].apply(lambda x: split_address(x))

0       (6649, N Blue Gum St, )
1      (4, B Blue Ridge Blvd, )
2      (8, W Cerritos Ave, #54)
3              (639, Main St, )
4             (34, Center St, )
                 ...           
495             (67, Rv Cent, )
496         (2, Sw Nyberg Rd, )
497        (89992, E 15th St, )
498       (61556, W 20th Ave, )
499         (63, E Aurora Dr, )
Name: address, Length: 500, dtype: object

Se quisermos criar novas colunas no DataFrame para cada parte do endereço, fazemos: (observe as últimas colunas)

In [36]:
try:
    us_500['address'] = us_500['address'].map(lambda x: split_address(x))
    # https://stackoverflow.com/questions/35491274/pandas-split-column-of-lists-into-multiple-columns
    us_500[['address_pt_1','address_pt_2', 'address_pt_3']] = pd.DataFrame(us_500['address'].tolist(), index=us_500.index)
    us_500 = us_500.drop(columns=['address'])
except:
    # Se rodar essa célula mais de uma vez, ela pode não funcionar pois o dataframe us_500 já foi alterado. 
    # Por isso apresento a seguinte mensagem:
    print("A operação não foi realizada. Possivelmente ela já tinha sido realizada anteriormente.")

us_500.head()

first_name last_name  ...       address_pt_2 address_pt_3
0      James      Butt  ...      N Blue Gum St             
1  Josephine   Darakjy  ...  B Blue Ridge Blvd             
2        Art    Venere  ...     W Cerritos Ave          #54
3      Lenna  Paprocki  ...            Main St             
4    Donette    Foller  ...          Center St             

[5 rows x 14 columns]